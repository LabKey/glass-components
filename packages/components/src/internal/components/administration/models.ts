interface Member {
    id: number;
    name: string;
    type: string;
}

interface Group {
    groupName: string;
    members: Member[];
}

export interface GroupMembership {
    [key: string]: Group;
}
