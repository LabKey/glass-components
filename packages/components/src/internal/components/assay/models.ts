/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { immerable, produce } from 'immer';

import { Operation } from '../../../public/QueryColumn';

import { AssayDefinitionModel } from '../../AssayDefinitionModel';
import { LoadingState } from '../../../public/LoadingState';

import { SelectInputChange } from '../forms/input/SelectInput';

import { AssayWizardModel } from './AssayWizardModel';

export interface AssayPropertiesPanelProps {
    operation: Operation;
    model: AssayWizardModel;
    onChange: (fieldValues: any, isChanged?: boolean) => void;
    onWorkflowTaskChange?: SelectInputChange;
}

export class AssayUploadResultModel {
    [immerable] = true;

    readonly assayId: number;
    readonly batchId: number;
    readonly runId: number;
    readonly success: boolean;
    readonly successurl?: string;
    readonly jobId?: number;

    constructor(values?: Partial<AssayUploadResultModel>) {
        Object.assign(this, values);
    }
}

export class AssayStateModel {
    [immerable] = true;

    readonly definitions: AssayDefinitionModel[];
    readonly definitionsError: string;
    readonly definitionsLoadingState: LoadingState;
    readonly protocolError: string;
    readonly protocolLoadingState: LoadingState;

    constructor(values?: Partial<AssayStateModel>) {
        Object.assign(this, values);

        this.definitions = this.definitions ?? [];
        this.definitionsLoadingState = this.definitionsLoadingState ?? LoadingState.INITIALIZED;
        this.protocolLoadingState = this.protocolLoadingState ?? LoadingState.INITIALIZED;
    }

    getById(assayRowId: number): AssayDefinitionModel {
        return this.definitions.find(def => def.id === assayRowId);
    }

    getByName(assayName: string): AssayDefinitionModel {
        const lowerName = assayName.toLowerCase();
        return this.definitions.find(def => def.name.toLowerCase() === lowerName);
    }

    getDefinitionsByTypes(included?: string[], excluded?: string[]): AssayDefinitionModel[] {
        if (!included && !excluded) return this.definitions;

        const lowerIncluded = included?.join('|').toLowerCase().split('|');
        const lowerExcluded = excluded?.join('|').toLowerCase().split('|');

        return this.definitions.filter(def => {
            let include = true;
            if (included?.length > 0) {
                include = lowerIncluded.indexOf(def.type.toLowerCase()) !== -1;
            }
            if (excluded?.length > 0) {
                include = lowerExcluded.indexOf(def.type.toLowerCase()) === -1;
            }
            return include;
        });
    }

    mutate(props: Partial<AssayStateModel>): AssayStateModel {
        return produce<AssayStateModel>(this, draft => {
            Object.assign(draft, props);
        });
    }
}
