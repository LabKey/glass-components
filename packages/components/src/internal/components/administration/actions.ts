import { List } from 'immutable'
import {User} from "../base/models/User";
import { ITab } from '../navigation/SubNav';
import {AppURL} from "../../url/AppURL";
import {SecurityPolicy} from "../permissions/models";
import {Security} from "@labkey/api";

export function getAdministrationSubNavTabs(user: User): List<ITab> {
    let tabs = List<string>();

    if (user.isAdmin) {
        tabs = tabs.push('Users');
        tabs = tabs.push('Permissions');
    }
    if (user.isAppAdmin()) {
        tabs = tabs.push('Settings');
    }

    return tabs.map(text => ({
        text,
        url: AppURL.create('admin', text.toLowerCase())
    })).toList()
}

export function getUserGridFilterURL(userIds: List<number>, urlPrefix: string): AppURL {
    let url = AppURL.create('admin', 'users');
    if (userIds && userIds.size > 0) {
        url = url.addParam(urlPrefix + '.UserId~in', userIds.join(';'));
    }
    return url;
}

export function updateSecurityPolicy(containerPath: string, userIds: List<number>, roleUniqueNames: string[]): Promise<any> {
    return new Promise((resolve, reject) => {
        Security.getPolicy({
            containerPath: containerPath,
            resourceId: containerPath,
            success: (data, relevantRoles) => {
                let policy = SecurityPolicy.create({policy: data, relevantRoles});
                userIds.forEach((userId) => {
                    roleUniqueNames.forEach(name => {
                        policy = SecurityPolicy.addUserIdAssignment(policy, userId, name);
                    });
                });

                Security.savePolicy({
                    containerPath: containerPath,
                    policy: {policy},
                    success: (response) => {
                        resolve(response);
                    },
                    failure: (error) => {
                        reject(error);
                    }
                });
            },
            failure: (error) => {
                reject(error);
            }
        });
    });
}
