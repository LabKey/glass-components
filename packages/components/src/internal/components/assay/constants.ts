import { SCHEMAS } from '../../schemas';

export const GENERAL_ASSAY_PROVIDER_NAME = 'General';

export const PLATE_METADATA_COLUMN = 'PlateMetadata';
export const PLATE_TEMPLATE_COLUMN = 'PlateTemplate';

// Columns are required for us to render the WorkflowTask in EditableDetails components
export const WORKFLOW_TASK_PROPERTIES_REQUIRED_COLUMNS = ['WorkflowTask', 'WorkflowTask/Run', 'Protocol/RowId'];

export const RUN_PROPERTIES_REQUIRED_COLUMNS = SCHEMAS.CBMB.concat(
    'Name',
    'RowId',
    'ReplacesRun',
    'ReplacedByRun',
    'DataOutputs',
    'DataOutputs/DataFileUrl',
    'Batch',
    'Folder'
)
    .concat(WORKFLOW_TASK_PROPERTIES_REQUIRED_COLUMNS)
    .toList();
