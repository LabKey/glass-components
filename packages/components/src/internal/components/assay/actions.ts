/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { List, Map, OrderedMap } from 'immutable';
import { Ajax, Assay, AssayDOM, Utils } from '@labkey/api';

import {
    AssayDefinitionModel,
    AssayStateModel,
    buildURL,
    caseInsensitive,
    naturalSortByProperty,
    SCHEMAS,
    QueryModel,
    User,
} from '../../..';

import { AssayUploadTabs } from '../../constants';

import { AssayUploadResultModel } from './models';
import { IAssayUploadOptions } from './AssayWizardModel';

export const GENERAL_ASSAY_PROVIDER_NAME = 'General';

export const RUN_PROPERTIES_REQUIRED_COLUMNS = SCHEMAS.CBMB.concat(
    'Name',
    'RowId',
    'ReplacesRun',
    'ReplacedByRun',
    'DataOutputs',
    'DataOutputs/DataFileUrl',
    'Batch',
    'Folder',
    // Below Columns are required for us to render the WorkflowTask in EditableDetails components
    'WorkflowTask',
    'WorkflowTask/Run',
    'Protocol/RowId'
).toList();

/**
 * Only support option to re-import run if user has insert permissions in current container
 * and the current container is where the source run is located. This prevents runs from being
 * re-imported "up" or "down" the folder structure.
 */
export function allowReimportAssayRun(user: User, runContainerId: string, targetContainerId: string): boolean {
    return (
        !!runContainerId && !!targetContainerId && targetContainerId === runContainerId && user.hasInsertPermission()
    );
}

let assayDefinitionCache: { [key: string]: Promise<List<AssayDefinitionModel>> } = {};

export function clearAssayDefinitionCache(): void {
    assayDefinitionCache = {};
}

export function fetchAllAssays(type?: string): Promise<List<AssayDefinitionModel>> {
    const key = type ?? 'undefined';

    if (!assayDefinitionCache[key]) {
        assayDefinitionCache[key] = new Promise((res, rej) => {
            Assay.getAll({
                parameters: { type },
                success: (rawModels: any[]) => {
                    const models = rawModels.reduce(
                        (list, raw) => list.push(AssayDefinitionModel.create(raw)),
                        List<AssayDefinitionModel>()
                    );
                    res(models);
                },
                failure: error => {
                    rej(error);
                },
            });
        });
    }

    return assayDefinitionCache[key];
}

export function importAssayRun(config: Partial<AssayDOM.IImportRunOptions>): Promise<AssayUploadResultModel> {
    return new Promise((resolve, reject) => {
        AssayDOM.importRun(
            Object.assign({}, config, {
                success: (rawModel: any) => {
                    resolve(new AssayUploadResultModel(rawModel));
                },
                failure: error => {
                    reject(error);
                },
            })
        );
    });
}

export function uploadAssayRunFiles(data: IAssayUploadOptions): Promise<IAssayUploadOptions> {
    return new Promise((resolve, reject) => {
        const batchFiles = collectFiles(data.batchProperties);
        const runFiles = collectFiles(data.properties);
        let maxFileSize = 0; // return the largest file size, used to determine if async mode should be used

        const maxRowCount = Array.isArray(data.dataRows) ? data.dataRows.length : undefined;
        if (data.files) {
            data.files.forEach(file => {
                if (file.size > maxFileSize) {
                    maxFileSize = file.size;
                }
            });
        }

        if (Utils.isEmptyObj(batchFiles) && Utils.isEmptyObj(runFiles)) {
            // No files in the data, so just go ahead and resolve so we run the import.
            resolve({ ...data, maxRowCount, maxFileSize });
            return;
        }

        // If we're this far along we've got files so let's process them.
        let fileCounter = 0;
        const formData = new FormData();
        const fileNameMap = {}; // Maps the file name "fileN" to the run/batch property it belongs to.

        // TODO factor this out so the code isn't duplicate below for the runFiles case
        Object.keys(batchFiles).forEach(columnName => {
            const name = fileCounter === 0 ? 'file' : `file${fileCounter}`;
            const file = batchFiles[columnName];
            if (file.size > maxFileSize) {
                maxFileSize = file.size;
            }
            fileNameMap[name] = {
                columnName,
                origin: 'batch',
            };
            formData.append(name, file);
            fileCounter++;
        });

        Object.keys(runFiles).forEach(columnName => {
            const name = fileCounter === 0 ? 'file' : `file${fileCounter}`;
            const file = runFiles[columnName];
            if (file.size > maxFileSize) {
                maxFileSize = file.size;
            }
            fileNameMap[name] = {
                columnName,
                origin: 'run',
            };
            formData.append(name, file);
            fileCounter++;
        });

        // N.B. assayFileUpload's success response is not handled well by Utils.getCallbackWrapper.
        Ajax.request({
            url: buildURL('assay', 'assayFileUpload.view'),
            method: 'POST',
            form: formData,
            success: result => {
                let response = JSON.parse(result.responseText);
                const batchPaths = {};
                const runPaths = {};

                if (fileCounter === 1) {
                    // Make the single file response look like the multi-file response.
                    response = {
                        file: response,
                    };
                }

                // Only process attributes that start with file so we ignore all the other stuff in the response like
                // the "success" attribute.
                Object.keys(response)
                    .filter(key => key.startsWith('file'))
                    .forEach(key => {
                        const { columnName, origin } = fileNameMap[key];

                        if (origin === 'batch') {
                            batchPaths[columnName] = response[key].absolutePath;
                        } else if (origin === 'run') {
                            runPaths[columnName] = response[key].absolutePath;
                        }
                    });

                resolve({
                    ...data,
                    batchProperties: {
                        ...data.batchProperties,
                        ...batchPaths,
                    },
                    properties: {
                        ...data.properties,
                        ...runPaths,
                    },
                    maxRowCount,
                    maxFileSize,
                });
            },
            failure: Utils.getCallbackWrapper(error => {
                // As far as I can tell errors returned from assay FileUpload are only ever strings.
                const message = `Error while uploading files: ${error}`;
                reject({ message });
            }),
        });
    });
}

interface FileMap {
    [s: string]: File;
}

function collectFiles(source: Record<string, any>): FileMap {
    return Object.keys(source).reduce((files, key) => {
        const item = source[key];

        if (item instanceof File) {
            files[key] = item;
        }

        return files;
    }, {} as FileMap);
}

export function deleteAssayRuns(
    selectionKey?: string,
    rowId?: string,
    cascadeDeleteReplacedRuns = false,
    containerPath?: string
): Promise<any> {
    return new Promise((resolve, reject) => {
        const jsonData: any = selectionKey ? { dataRegionSelectionKey: selectionKey } : { singleObjectRowId: rowId };
        jsonData.cascade = cascadeDeleteReplacedRuns;

        return Ajax.request({
            url: buildURL('experiment', 'deleteRuns.api', undefined, {
                container: containerPath,
            }),
            method: 'POST',
            jsonData,
            success: Utils.getCallbackWrapper(response => {
                resolve(response);
            }),
            failure: Utils.getCallbackWrapper(response => {
                reject(response);
            }),
        });
    });
}

export function getImportItemsForAssayDefinitions(
    assayStateModel: AssayStateModel,
    sampleModel?: QueryModel,
    providerType?: string
): OrderedMap<AssayDefinitionModel, string> {
    let targetSQ;
    const selectionKey = sampleModel?.id;

    if (sampleModel?.queryInfo) {
        targetSQ = sampleModel.queryInfo.schemaQuery;
    }

    return assayStateModel.definitions
        .filter(assay => providerType === undefined || assay.type === providerType)
        .filter(assay => !targetSQ || assay.hasLookup(targetSQ))
        .sort(naturalSortByProperty('name'))
        .reduce((items, assay) => {
            const href = assay.getImportUrl(
                selectionKey ? AssayUploadTabs.Grid : AssayUploadTabs.Files,
                selectionKey,
                // Check for the existence of the "queryInfo" before getting filters from the model.
                // This avoids `QueryModel` throwing an error when the "queryInfo" is not yet available.
                sampleModel?.queryInfo ? List(sampleModel.filters) : undefined
            );
            return items.set(assay, href);
        }, OrderedMap<AssayDefinitionModel, string>());
}

export interface DuplicateFilesResponse {
    duplicate: boolean;
    newFileNames: List<string>;
    runNamesPerFile: List<string>;
}

export function checkForDuplicateAssayFiles(fileNames: string[]): Promise<DuplicateFilesResponse> {
    return new Promise((resolve, reject) => {
        Ajax.request({
            url: buildURL('assay', 'assayFileDuplicateCheck.api'),
            method: 'POST',
            jsonData: {
                fileNames,
            },
            success: Utils.getCallbackWrapper(res => {
                resolve(res);
            }),
            failure: Utils.getCallbackWrapper(response => {
                console.error('Problem checking for duplicate files', response);
                reject(response);
            }),
        });
    });
}

export function getRunPropertiesFileName(row: Record<string, any>): string {
    const dataOutputs = caseInsensitive(row, 'DataOutputs');
    return dataOutputs && dataOutputs.length === 1 ? dataOutputs[0].displayValue : undefined;
}

export function flattenQueryModelRow(rowData: Record<string, any>): Map<string, any> {
    if (rowData) {
        // TODO make the consumers of this row data able to handle the queryData instead of
        // having to create the key -> value map via reduction.
        let map = Map<string, any>();

        Object.keys(rowData).forEach(k => {
            const v = rowData[k];
            let valueMap = v;
            if (Utils.isArray(v)) {
                if (v.length > 1) {
                    console.warn("Multiple values for field '" + k + "'.  Using the last.");
                }
                valueMap = v[v.length - 1];
            }
            if (valueMap && valueMap['value'] !== undefined && valueMap['value'] !== null) {
                map = map.set(k, valueMap['value']);
            }
        });

        return map;
    }

    return Map<string, any>();
}

export function deleteAssayDesign(rowId: string): Promise<any> {
    return new Promise((resolve, reject) => {
        return Ajax.request({
            url: buildURL('experiment', 'deleteProtocolByRowIdsAPI.api'),
            method: 'POST',
            params: {
                singleObjectRowId: rowId,
                forceDelete: true,
            },
            success: Utils.getCallbackWrapper(response => {
                resolve(response);
            }),
            failure: Utils.getCallbackWrapper(response => {
                reject(response);
            }),
        });
    });
}
