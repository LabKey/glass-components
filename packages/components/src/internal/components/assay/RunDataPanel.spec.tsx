import React from 'react';
import renderer from 'react-test-renderer';

import { getQueryGridModel } from '../../global';
import { getStateQueryGridModel } from '../../models';
import { gridInit } from '../../actions';
import { withFormSteps, WithFormStepsProps } from '../forms/FormStep';
import { ASSAY_WIZARD_MODEL } from '../../../test/data/constants';
import { initUnitTestMocks } from '../../testHelpers';
import { AssayUploadTabs } from '../base/models/model';

import { RunDataPanel } from './RunDataPanel';

let MODEL_ID_NOT_LOADED;
let MODEL_ID_LOADED;

beforeAll(() => {
    initUnitTestMocks();

    let model = getStateQueryGridModel('jest-test-0', ASSAY_WIZARD_MODEL.queryInfo.schemaQuery, {
        editable: true,
        allowSelection: false,
        bindURL: false,
    });
    MODEL_ID_NOT_LOADED = model.getId();

    model = getStateQueryGridModel('jest-test-1', ASSAY_WIZARD_MODEL.queryInfo.schemaQuery, {
        editable: true,
        allowSelection: false,
        bindURL: false,
    });

    gridInit(model, false);
    MODEL_ID_LOADED = model.getId();
});

interface OwnProps {
    fullWidth?: boolean;
    allowBulkRemove?: boolean;
}
type Props = OwnProps & WithFormStepsProps;

class RunDataPanelWrapperImpl extends React.Component<Props, any> {
    render() {
        const { currentStep, fullWidth, allowBulkRemove } = this.props;
        const gridModel = getQueryGridModel(MODEL_ID_LOADED);

        return (
            <RunDataPanel
                currentStep={currentStep}
                wizardModel={ASSAY_WIZARD_MODEL}
                gridModel={gridModel}
                onFileChange={jest.fn}
                onFileRemoval={jest.fn}
                onTextChange={jest.fn}
                fullWidth={fullWidth}
                allowBulkRemove={allowBulkRemove}
            />
        );
    }
}

const RunDataPanelWrapper = withFormSteps(RunDataPanelWrapperImpl, {
    currentStep: AssayUploadTabs.Files,
    furthestStep: AssayUploadTabs.Grid,
    hasDependentSteps: false,
});

describe('<RunDataPanel/>', () => {
    test('loading state based on gridModel', () => {
        const component = (
            <RunDataPanel
                currentStep={AssayUploadTabs.Files}
                wizardModel={ASSAY_WIZARD_MODEL}
                gridModel={getQueryGridModel(MODEL_ID_NOT_LOADED)}
                onFileChange={jest.fn}
                onFileRemoval={jest.fn}
                onTextChange={jest.fn}
            />
        );

        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('default props', () => {
        const component = <RunDataPanelWrapper />;

        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('custom display props', () => {
        const component = <RunDataPanelWrapper allowBulkRemove={true} fullWidth={true} />;

        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
