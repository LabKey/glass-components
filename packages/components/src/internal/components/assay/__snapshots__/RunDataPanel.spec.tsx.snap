// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RunDataPanel/> custom display props 1`] = `
<div
  className="panel panel-default full-width"
>
  <div
    className="panel-heading"
  >
    Results
  </div>
  <div
    className="panel-body"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-12"
      >
        <ul
          className="list-group clearfix"
          style={
            Object {
              "listStyle": "none",
            }
          }
        >
          <li
            className="list-group-item form-step-tab active"
            onClick={[Function]}
          >
            Upload Files
          </li>
          <li
            className="list-group-item form-step-tab"
            onClick={[Function]}
          >
            Copy-and-Paste Data
          </li>
          <li
            className="list-group-item form-step-tab"
            onClick={[Function]}
          >
            Enter Data Into Grid
          </li>
        </ul>
        <div
          className="clearfix"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-sm-12"
      >
        <div
          className="form-step active"
        >
          <span
            className="translator--toggle__wizard"
          >
            <div
              className="row"
            >
              <div
                className="col-sm-12"
              >
                <div
                  className="wizard-row--container"
                >
                  <div
                    className=""
                  >
                    <div
                      className="file-upload--container block"
                    >
                      <label
                        className="file-upload--label"
                        htmlFor="fileUpload"
                        onDragEnter={[Function]}
                        onDragLeave={[Function]}
                        onDragOver={[Function]}
                        onDrop={[Function]}
                      >
                        <i
                          aria-hidden="true"
                          className="fa fa-cloud-upload fa-2x cloud-logo"
                        />
                        Select file or drag and drop here.
                      </label>
                      <input
                        accept=""
                        className="file-upload--input"
                        id="fileUpload"
                        multiple={false}
                        name="fileUpload"
                        onChange={[Function]}
                        type="file"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </span>
          <div
            className="row"
          >
            <div
              className="col-md-9"
            >
              
            </div>
            <div
              className="col-md-3"
            >
              <div
                className="pull-right"
              >
                <a
                  className="btn btn-info"
                  href="/labkey/Sample%20Management%202/assay-template.view?rowId=5051"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  title="Download Template"
                >
                  <span
                    className="fa fa-download"
                  />
                   Template
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="form-step"
        >
          <form
            disabled={false}
            onError={[Function]}
            onReset={[Function]}
            onSubmit={[Function]}
          >
            <div
              className="form-group row"
            >
              <label
                className="control-label hidden"
                data-required={false}
                htmlFor="frc-rundata-1088"
              >
                
              </label>
              <div
                className="col-sm-12"
              >
                <textarea
                  className="form-control"
                  cols={-1}
                  disabled={false}
                  id="frc-rundata-1088"
                  name="rundata"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onKeyDown={[Function]}
                  placeholder="Paste in a tab-separated set of values (including column headers)."
                  required={false}
                  rows={10}
                  value=""
                />
              </div>
            </div>
            <button
              className="btn btn-default"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Clear
            </button>
          </form>
        </div>
        <div
          className="form-step"
        >
          <div>
            <div
              className="row QueryGrid-bottom-spacing"
            >
              <div
                className="col-sm-3"
              >
                <div
                  className="editable-grid__controls text-nowrap"
                >
                  <span
                    className="input-group input-group-align"
                  >
                    <input
                      className="form-control"
                      disabled={false}
                      max={1000}
                      min={1}
                      name="addCount"
                      onBlur={[Function]}
                      onChange={[Function]}
                      style={
                        Object {
                          "width": "65px",
                        }
                      }
                      type="number"
                      value="1"
                    />
                    <span
                      className="input-group-btn"
                    >
                      <button
                        className="btn btn-primary"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        Add row
                      </button>
                    </span>
                  </span>
                </div>
              </div>
              <div
                className="col-sm-9"
              >
                <span
                  className="control-right"
                >
                  <button
                    className="control-right btn btn-default"
                    disabled={true}
                    onClick={[Function]}
                    type="button"
                  >
                    Delete Rows
                  </button>
                </span>
              </div>
            </div>
            <div
              className="editable-grid__container"
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
            >
              <div
                className=""
              >
                <div
                  className="grid-messages"
                />
                <table
                  className="table-cellular table-striped table-bordered"
                >
                  <thead>
                    <tr>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": undefined,
                          }
                        }
                      >
                        <input
                          checked={false}
                          disabled={false}
                          onChange={[Function]}
                          type="checkbox"
                        />
                      </th>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": "45px",
                          }
                        }
                      >
                        Row
                      </th>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": "100px",
                          }
                        }
                      >
                        Specimen ID
                      </th>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": "100px",
                          }
                        }
                      >
                        Participant ID
                      </th>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": "100px",
                          }
                        }
                      >
                        Visit ID
                      </th>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": "100px",
                          }
                        }
                      >
                        Date
                      </th>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": "100px",
                          }
                        }
                      >
                        Healthy
                      </th>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": "100px",
                          }
                        }
                      >
                        File
                      </th>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": "100px",
                          }
                        }
                      >
                        Text Area
                      </th>
                    </tr>
                  </thead>
                  <tbody>
                    <tr
                      className="grid-empty"
                    >
                      <td
                        colSpan={9}
                      >
                        Start by adding the quantity of assay data rows you want to create.
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="top-spacing"
    />
  </div>
</div>
`;

exports[`<RunDataPanel/> default props 1`] = `
<div
  className="panel panel-default full-width"
>
  <div
    className="panel-heading"
  >
    Results
  </div>
  <div
    className="panel-body"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-12"
      >
        <ul
          className="list-group clearfix"
          style={
            Object {
              "listStyle": "none",
            }
          }
        >
          <li
            className="list-group-item form-step-tab active"
            onClick={[Function]}
          >
            Upload Files
          </li>
          <li
            className="list-group-item form-step-tab"
            onClick={[Function]}
          >
            Copy-and-Paste Data
          </li>
          <li
            className="list-group-item form-step-tab"
            onClick={[Function]}
          >
            Enter Data Into Grid
          </li>
        </ul>
        <div
          className="clearfix"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-sm-12"
      >
        <div
          className="form-step active"
        >
          <span
            className="translator--toggle__wizard"
          >
            <div
              className="row"
            >
              <div
                className="col-sm-12"
              >
                <div
                  className="wizard-row--container"
                >
                  <div
                    className=""
                  >
                    <div
                      className="file-upload--container block"
                    >
                      <label
                        className="file-upload--label"
                        htmlFor="fileUpload"
                        onDragEnter={[Function]}
                        onDragLeave={[Function]}
                        onDragOver={[Function]}
                        onDrop={[Function]}
                      >
                        <i
                          aria-hidden="true"
                          className="fa fa-cloud-upload fa-2x cloud-logo"
                        />
                        Select file or drag and drop here.
                      </label>
                      <input
                        accept=""
                        className="file-upload--input"
                        id="fileUpload"
                        multiple={false}
                        name="fileUpload"
                        onChange={[Function]}
                        type="file"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </span>
          <div
            className="row"
          >
            <div
              className="col-md-9"
            >
              
            </div>
            <div
              className="col-md-3"
            >
              <div
                className="pull-right"
              >
                <a
                  className="btn btn-info"
                  href="/labkey/Sample%20Management%202/assay-template.view?rowId=5051"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  title="Download Template"
                >
                  <span
                    className="fa fa-download"
                  />
                   Template
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="form-step"
        >
          <form
            disabled={false}
            onError={[Function]}
            onReset={[Function]}
            onSubmit={[Function]}
          >
            <div
              className="form-group row"
            >
              <label
                className="control-label hidden"
                data-required={false}
                htmlFor="frc-rundata-1088"
              >
                
              </label>
              <div
                className="col-sm-12"
              >
                <textarea
                  className="form-control"
                  cols={-1}
                  disabled={false}
                  id="frc-rundata-1088"
                  name="rundata"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onKeyDown={[Function]}
                  placeholder="Paste in a tab-separated set of values (including column headers)."
                  required={false}
                  rows={10}
                  value=""
                />
              </div>
            </div>
            <button
              className="btn btn-default"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Clear
            </button>
          </form>
        </div>
        <div
          className="form-step"
        >
          <div>
            <div
              className="row QueryGrid-bottom-spacing"
            >
              <div
                className="col-sm-3"
              >
                <div
                  className="editable-grid__controls text-nowrap"
                >
                  <span
                    className="input-group input-group-align"
                  >
                    <input
                      className="form-control"
                      disabled={false}
                      max={1000}
                      min={1}
                      name="addCount"
                      onBlur={[Function]}
                      onChange={[Function]}
                      style={
                        Object {
                          "width": "65px",
                        }
                      }
                      type="number"
                      value="1"
                    />
                    <span
                      className="input-group-btn"
                    >
                      <button
                        className="btn btn-primary"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        Add row
                      </button>
                    </span>
                  </span>
                </div>
              </div>
              <div
                className="col-sm-9"
              />
            </div>
            <div
              className="editable-grid__container"
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
            >
              <div
                className=""
              >
                <div
                  className="grid-messages"
                />
                <table
                  className="table-cellular table-striped table-bordered"
                >
                  <thead>
                    <tr>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": "45px",
                          }
                        }
                      >
                        Row
                      </th>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": "100px",
                          }
                        }
                      >
                        Specimen ID
                      </th>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": "100px",
                          }
                        }
                      >
                        Participant ID
                      </th>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": "100px",
                          }
                        }
                      >
                        Visit ID
                      </th>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": "100px",
                          }
                        }
                      >
                        Date
                      </th>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": "100px",
                          }
                        }
                      >
                        Healthy
                      </th>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": "100px",
                          }
                        }
                      >
                        File
                      </th>
                      <th
                        className="grid-header-cell"
                        onClick={[Function]}
                        style={
                          Object {
                            "minWidth": "100px",
                          }
                        }
                      >
                        Text Area
                      </th>
                    </tr>
                  </thead>
                  <tbody>
                    <tr
                      className="grid-empty"
                    >
                      <td
                        colSpan={8}
                      >
                        Start by adding the quantity of assay data rows you want to create.
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="top-spacing"
    />
  </div>
</div>
`;

exports[`<RunDataPanel/> loading state based on gridModel 1`] = `
<div
  className="panel panel-default full-width"
>
  <div
    className="panel-heading"
  >
    Results
  </div>
  <div
    className="panel-body"
  >
    <span
      className=""
    >
      <i
        aria-hidden="true"
        className="fa fa-spinner fa-pulse"
      />
       
      Loading...
    </span>
  </div>
</div>
`;
